### PR TITLE
Added Fix and tests for calling WithImage (#2580)

### LINKED ISSUE #2580: WithImage and WithImageTag assume a single ContainerImageAnnoation exists
The `WithImage` and `WithImageTag` methods assume the `ContainerResource` has a single `ContainerImageAnnotation` but by default they have none, and they could have more than one. Calling either of these methods in these states results in a LINQ exception.

We should make these methods more user friendly, e.g. add the `ContainerImageAnnotation` if it doesn't exist. If there's more than one then perhaps it should just get the last one and update that.

### PR DESCRIPTION
**Pull request by GMouaad**
When calling `WithImage`, we now first check if the annotation exist and mutate its image. if it doesn't we add the annotation with the given image.

Fixes #2580

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2588)

### CODE REVIEW

**Comment by GMouaad** [comment]
@microsoft-github-policy-service agree

**Comment by davidfowl** [comment]
This change looks pretty good. Thoughts @mitchdenny ? Should we do this for the other container helper methods as well?

**Comment by GMouaad** [comment]
@dotnet-policy-bot agree

**Comment by DamianEdwards** [comment]
@GMouaad you need to address your comment to *@dotnet-policy-service*, not *@dotnet-policy-bot*.

**Comment by GMouaad** [comment]
@dotnet-policy-service agree

**Comment by davidfowl** [comment]
Is it valid to add an image without a tag? Do we default to latest? I'm not convinced these methods work well since they independently mutate pieces of the same annotation which allow putting them into an inconsistent state.

**Comment by GMouaad** [comment]
@davidfowl I agree.
> Do we default to latest?

I don't think we ensure that everywhere. The `AddContainer()` without the `tag` overload does that. Though there are places where the `Tag` is not _required_ and may slip through as empty. 

@mitchdenny , do you think that removing the `WithImageTag()` method and adding an overload to `WithImage(string image, string tag)` would be a better approach and more usable? or any other suggestions?
```diff

public static class ContainerResourceBuilderExtensions
{
 public static IResourceBuilder<T> WithImage<T>(this IResourceBuilder<T> builder, string image) where T : ContainerResource // => WithImage<T>(image, "latest")

+ public static IResourceBuilder<T> WithImage<T>(this IResourceBuilder<T> builder, string image, string tag) where T : ContainerResource

- public static IResourceBuilder<T> WithImageTag<T>(this IResourceBuilder<T> builder, string tag) where T : ContainerResource
}

```
If yes, I could incorporate that in this PR.

**Comment by davidfowl** [comment]
- We want to allow people to override the tag without the image (change from a version to latest)
- We want people to pick the image and tag (we don't want to assume latest)

If you try to set the tag without already having set the image, it should throw.
If you try to override the image, then you should be forced to pass the tag as well. Just like AddContainer does today.

File: src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs [hunk 1]
```diff
@@ -129,12 +133,33 @@ public static IResourceBuilder<ContainerResource> AddContainer(this IDistributed
...
+    {
+        if (builder.Resource.Annotations.OfType<ContainerImageAnnotation>().LastOrDefault() is { } existingImageAnnotation)
+        {
+            existingImageAnnotation.Image = image;
+            existingImageAnnotation.Tag = tag;
+            return builder;
+        }
+
+        // if the annotation doesn't exist, create it with the given image and add it to the collection
+        var containerImageAnnotation = new ContainerImageAnnotation() { Image = image };
```
Comment (davidfowl) [inline src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs:161]: and tag?

File: src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs [hunk 2]
```diff
@@ -105,9 +105,13 @@ public static IResourceBuilder<ContainerResource> AddContainer(this IDistributed
...
-        var containerImageAnnotation = builder.Resource.Annotations.OfType<ContainerImageAnnotation>().Single();
-        containerImageAnnotation.Tag = tag;
-        return builder;
+        if (builder.Resource.Annotations.OfType<ContainerImageAnnotation>().LastOrDefault() is { } existingImageAnnotation)
+        {
+            existingImageAnnotation.Tag = tag;
+            return builder;
+        }
+
+        throw new InvalidOperationException("The resource does not contain any image annotation to be mutated");
```
Comment (davidfowl) [inline src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs:114]:
```suggestion
        throw new InvalidOperationException($"The resource '{builder.Resource.Name}' does not have a container image specified. Use WithImage to specify the container image and tag.");
```

File: src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs [hunk 3]
```diff
@@ -129,12 +133,33 @@ public static IResourceBuilder<ContainerResource> AddContainer(this IDistributed
...
+
+    /// <summary>
+    /// Allows overriding the image on a container.
+    /// </summary>
+    /// <typeparam name="T">Type of container resource.</typeparam>
+    /// <param name="builder">Builder for the container resource.</param>
+    /// <param name="image">Image value.</param>
+    /// <param name="tag">Tag value.</param>
+    /// <returns></returns>
+    public static IResourceBuilder<T> WithImage<T>(this IResourceBuilder<T> builder, string image, string tag) where T : ContainerResource
```
Comment (davidfowl) [inline src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs:151]: Do we need 2 overloads? Can't we just do tag= "latest" as a default argument?

File: src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs (hunk 1 above)
Comment (GMouaad) [reply src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs]: Missed that one :) good spot

File: src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs (hunk 3 above)
Comment (GMouaad) [reply src/Aspire.Hosting/Extensions/ContainerResourceBuilderExtensions.cs:151]: We surely can, I just tried to follow the same pattern used in the other extensions

**Comment by davidfowl** [comment]
Looks good.

**Review by davidfowl — APPROVED** [review]

**Comment by davidfowl** [comment]
Thanks @GMouaad !

**Comment by GMouaad** [comment]
Glad to contribute! Thanks for guiding through my first issue here :). 
I'll see if I can get my hands on other issues ;)

**Pull request merged**